Feature/aws containers

ggotti opened the pull request:
This PR adds a new AWS specific sub-generator, called 'aws-containers', to deploy JHhipster applications to Amazon using Elastic Container Service (ECS). The version in this PR only works with monolithic applications. At a high level it does the following:
- 	Queries your AWS account for details around deployment topology: VPCs, subnets, database passwords.
- Introduces SpringCloud dependencies to allow your application to query Amazon's System Manager for credentials on startup.
- Uses a nested-stack CloudFormation arrangement to deploy your selected applications. 
- Creates an Elastic Container Registry (ECR) or private Docker repository to deploy applications to.

A detailed update of the various AWS technologies and services can be seen in this website PR. 

Thanks to @carvallegro for doing most of the work :)

Addresses monolithic component of #6773 
- [ ] [Travis tests](https://travis-ci.org/jhipster/generator-jhipster/pull_requests) are green
- [ ] Tests are added where necessary
- [X] Documentation is added/updated where necessary
- [X] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/master/CONTRIBUTING.md) are followed

deepu105 commented:
Lets do this for v5 as well since its a major feature

deepu105 commented:
@jhipster/developers who can review this? I can do the JS part but not familiar with AWS containers

MathieuAA commented on generators/aws-containers/README.md:
```diff
@@ -0,0 +1,24 @@
+## AWS
+### ECR
+#### Authentication Token
+The token received by ECR to authenticate is structured as follow. We'll only take the first object.
```
`follow` -> `followed`

carvallegro replied:
It's actually 'as follows' but thanks for the catch up :)

MathieuAA replied:
Right!! thanks

MathieuAA commented on generators/aws-containers/README.md:
````diff
@@ -0,0 +1,24 @@
+
+```
+# NPM
+npm i --save-dev aws-sdk@^2.167.0 progress@2.0.0 ora@1.3.0
````
Can the versions be fixed for both yarn and npm?

carvallegro replied:
I'm not sure what you mean here. The installation instructions for both yarn and npm is the same

deepu105 replied:
can we put the version numbers in a constant and use that in both places so that at least you only need to change it once

carvallegro replied:
It is not possible to use variables in Markdown files but we could put constants in `aws-client.js` and maybe put a TODO comment to remind to update the doc

MathieuAA replied:
By fixed I meant without `^` in the versions, like `npm i --save-dev aws-sdk@2.167.0 progress@2.0.0 ora@1.3.0`

deepu105 replied:
Oh sorry didn't notice that this was the readme of the sub gen itself

MathieuAA commented on generators/aws-containers/aws-client.js:
```diff
@@ -0,0 +1,299 @@
+            generator.log('Installing AWS dependencies');
+            let installCommand = 'yarn add aws-sdk@^2.167.0 progress@2.0.0 ora@1.3.0';
+            if (generator.config.get('clientPackageManager') === 'npm') {
+                installCommand = 'npm install aws-sdk@^2.167.0 progress@2.0.0 ora@1.3.0--save';
```
If the version changes we'd have to update it here too... is there a better solution?

carvallegro replied:
We'll definitely look into it but the issue with not specifying the version is that we don't have any control over the installed version. The code works with `2.167.0` so we're sure to avoid any breaking changes.

For example if `aws-sdk` has a 3.0.0 major release it might break the code. Hardcoding a version means that we only have to be careful at the npm version required by the `jhipster-gernerator` and update accordinlgy, which will happens less often than a `aws-sdk`update

MathieuAA commented on generators/aws-containers/aws-client.js:
```diff
@@ -0,0 +1,299 @@
+}
+
+/**
+ * listVpcs(region) returns a Promise, which resolves to an array of AWS VPC objects
```
The doc is not consistent with the code here

MathieuAA commented on generators/aws-containers/aws-client.js:
```diff
@@ -0,0 +1,299 @@
+/**
+ * listSubnets(vpcId) returns a Promise, which resolves to an array of
+ * Subnets available within the supplied VPC
+ * @param region to look for subnets
```
I guess you forgot to remove the param

MathieuAA commented on generators/aws-containers/aws-client.js:
```diff
@@ -0,0 +1,299 @@
+        .then(authToken => sts.getCallerIdentity({}).promise()
+            .then((data) => {
+                const decoded = utils.decodeBase64(authToken.authorizationToken);
+                const splitResult = _.split(decoded, ':');
```
lodash's `split` can be replaced by a regular `split`

MathieuAA commented on generators/aws-containers/aws-client.js:
```diff
@@ -0,0 +1,299 @@
+                reject(new Error(`The S3 Bucket ${chalk.bold(bucketName)} in region ${chalk.bold(region)} already exists and you don't have access to it. Error code: ${chalk.bold(error.code)}`));
+            }
+        })
+        .then(() =>
```
Can you call `then` before `catch` please?

carvallegro replied:
It was originally like this but because of the way Promises behave we've had to put it first. If we put the `catch` `after` the `then` and the `s3.createBucket()` promise fails, it falls back to the `spinner()` which will change the expected behavior

MathieuAA replied:
That's weird.

jdubois commented:
Biggest issue is that you need AWS $$$$ for that

MathieuAA commented on generators/aws-containers/aws-client.js:
```diff
@@ -0,0 +1,299 @@
+function uploadTemplate(bucketName, filename, path) {
+    return spinner(new Promise((resolve, reject) =>
+        fs.stat(path, (error, stats) => {
+            if (_.isNil(stats)) {
```
As lodash's `isNil` just does `value == null`, I think you can just write `!stats` here

MathieuAA commented on generators/aws-containers/aws-client.js:
```diff
@@ -0,0 +1,299 @@
+            );
+            let bar;
+            upload.on('httpUploadProgress', (evt) => {
+                if (bar === undefined && evt.total) {
```
`!bar`

MathieuAA commented on generators/aws-containers/aws-client.js:
```diff
@@ -0,0 +1,299 @@
+ * @returns {string}
+ */
+function sanitizeBucketName(bucketName) {
+    let labels = _.split(bucketName, '.');
```
Two things:
  - Instead of using lodash, you can use the regular API
  - You can chain the calls

MathieuAA commented on generators/aws-containers/lib/cf.js:
```diff
@@ -0,0 +1,297 @@
+
+let stdOut = message => console.error(message.trim()); // eslint-disable-line
+let stdErr = message => console.error(message.trim()); // eslint-disable-line
+module.exports = class CF {
```
Is it okay to rename the class to `CloudFormation ` instead of using an abbreviation?

carvallegro replied:
We've named it `cf` in order to be consistant with the other classes in `lib/` : `ecr` and `ssm`

MathieuAA commented on generators/aws-containers/lib/cf.js:
```diff
@@ -0,0 +1,297 @@
+let stdErr = message => console.error(message.trim()); // eslint-disable-line
+module.exports = class CF {
+    constructor(region) {
+        this.cloudFormation = new AWS.CloudFormation({ region });
```
If `CF` means `CloudFormation`, is the parameter an internal object?

carvallegro replied:
I'm not sure what you mean but here the constructor parameter, `region` is used to build `this.cloudFormation` which is an AWS component. It's a composition pattern.

MathieuAA replied:
I know. What I meant is I'd rather change the file's name (and the class' name too) and rename the `couldFormation` field.

carvallegro commented:
@MathieuAA I've pushed some changes around your comments

MathieuAA commented on generators/generator-base.js:
```diff
@@ -1447,7 +1460,19 @@ module.exports = class extends PrivateBase {
+     * @param {string} version - (optional) explicit dependency version number
+     */
+    addGradleDependencyInDirectory(directory, scope, group, name, version) {
+        const fullPath = `${directory}/build.gradle`;
```
@deepu105 Maybe the use of the `path` module would be good here, don't you think?

MathieuAA commented on generators/docker-utils.js:
```diff
@@ -0,0 +1,96 @@
+    checkImageExist.call(this, opts);
+    const pwd = shelljs.pwd();
+    shelljs.cd(opts.cwd);
+    return new Promise((resolve, reject) => dockerCLI.command(`${opts.cwd}${this.dockerBuildCommand}`, (err) => {
```
Can you format here?
```javascript
return new Promise((resolve, reject) =>
  dockerCLI.command(`${opts.cwd}${this.dockerBuildCommand}`, (err) => {
```

jdubois commented:
I will do a review of this code on Tuesday -> @ggotti could you also prepare some documentation for this?
Something like http://www.jhipster.tech/aws/ (the source is at https://github.com/jhipster/jhipster.github.io - and don't worry if this is not perfect graphically etc, what I need is documentation, I'll format and correct it myself)

jdubois commented:
Please have a look at my reviews, and update jhipster-dependencies.

Now I can deploy, but I have 2 major issues:

- My database connection fails, I got ` SQLException occurred while connecting to jhipster-test-ecs-tmp4-1ob0zht-jhipsterrdscluster-1lw3q3us4xyr7.cluster-co7tgvzcrcxe.us-east-1.rds.amazonaws.com:3306`- this is with Aurora/Postgresql. Please note I can see the database in Amazon RDS, so it's a configuration issue on our side.
- I'm worried that this created lots of services, and I have no idea how to clean up what was created. Can we have some kind of "tag" somewhere, to know what was deployed, how much does it costs, and delete everything easily?

jdubois commented:
You will also need to update the limitations in the documentation:

- this doesn't work with the JHipster Registry.
- if you use a cache, it will not scale (because no JHipster Registry), so it's easier a non-scalable ehcache cache, or no cache. That's a bit annoying.
- this only supports SQL databases with Aurora -> no MongoDB, Cassandra, etc.... so this should explained upfront

-> what would be even better is to check this when the sub-generator runs, and make it fail with a nice error message if there's something wrong

ggotti commented:
I'll look at actioning the PR feedback early next week, as I'm currently studying for an exam. In terms of comments in this thread. @jdubois 
-I'll look at adding tagging support in to the components where possible. However it should be noted that you should use the CloudFormation delete functionality where possible, as it makes it much easier to cleanup.
- There is an associated documentation PR [here](https://github.com/jhipster/jhipster.github.io/pull/532).
- I'll look at adding a pre-check in for the database configuration for ensure that it only creates for valid database configurations.

carvallegro commented on generators/aws-containers/templates/_base.template.yml:
```diff
@@ -0,0 +1,35 @@
+    Properties:
+      Parameters:
+        parentStackName: !Ref AWS::StackName
+        repositoryName: <%= app.baseName %>
```
@ggotti You can't have uppercase in the repository name. Needs to comply with the following regex:
```
 '(?:[a-z0-9]+(?:[._-][a-z0-9]+)*/)*[a-z0-9]+(?:[._-][a-z0-9]+)*'
```

ggotti commented:
Thanks to everyone for the feedback. We've made a bunch of changes based on the feedback we have received:

- 	Corrected issues with the Postgresql deployment option.
- 	Split out performance from scaling.
- 	Renamed templates to have a ejs suffix.
- 	Replaced Network Load Balancer with Application Load Balancer.
- 	Added a warning about distributed caching.
- 	Tightening of input control around stackname+application names.

The documentation [PR](https://github.com/jhipster/jhipster.github.io/pull/532)) has also been updated to reflect these changes.

jdubois commented:
@ggotti I tested it twice, I consistently got:

```
12:14:54 AM 	CREATE FAILED                       AWS::RDS::DBClusterParameterGroup	jhipster-test-ippon-tmp-ER0MYQ8VWR4X
12:14:55 AM 	CREATE FAILED                       AWS::CloudFormation::Stack	jhipster-test-ippon-tmp-ER0MYQ8VWR4X
✖ There was an error creating the stack: Creation of nested stack failed
```

Could you have a look?

ggotti commented:
Ok so the issue that @jdubois has raised is an interesting one, and was caused by insufficient timeout given to CloudFormation for the stack. The sub-generator was creating a "AWS::RDS::DBClusterParameterGroup", which according to the [CLI documentation](https://docs.aws.amazon.com/cli/latest/reference/rds/create-db-cluster-parameter-group.html) (but interestingly enough not the CloudFormation documentation) requires at least 5 minutes before creating your new DB Cluster. We had set a 5 minute timeout on the nested-stack creation itself, and so it appears that we're now hitting that timeout limit; the sub-stack timeout is for individual items, not the stack itself. 

So, where to from here? The DBClusterParameterGroup is actually not giving us a lot of value, and speed wise is more of a headache than its worth. So what I've done is removed it from the CloudFormation nest-stack, and used the account default values (as per [this](https://docs.aws.amazon.com/AmazonRDS/latest/APIReference/API_CreateDBCluster.html)).  I've also removed the timeout function, which will account for situations where AWS is taking longer to provision than normal.

jdubois commented:
OK I tested it several times yesterday without any issue, let's merge this!
